File: pat1102.cpp
/*
	思路：1.用结构体数组存储输入数据
		  2.由左右节点没出现的节点为根节点的原理，找出根节点
		  3.通过层序反转遍历，中序反转遍历输出结果

*/


#include<iostream>
#include<queue>
using namespace std;

struct node{
	int left = -1;
	int right = -1;
}Tree[10];

//	将输入的char类型左右子节点，转化为int型，无子节点则赋值为-1
int transfer(char c) {
	if (c == '-') {
		return -1;
	}
	else {
		return c - '0';
	}
}

//	层序反转遍历
void level_invert_order(int root) {
	queue<int> q;
	q.push(root);
	cout << root;
	while (!q.empty()){
		int pop_node = q.front();
		int l = Tree[pop_node].left;
		int r = Tree[pop_node].right;
		if (r != -1) {
			cout << " " << r;
			q.push(r);
		}
		if (l != -1) {
			cout << " " << l;
			q.push(l);
		}
		q.pop();
	}
}

//	中序反转遍历
bool first_out = true;
void inorder_invert(int root) {
	if (root == -1) return;
	inorder_invert(Tree[root].right);
	//	第一次输出和后面几次输出有一些区别
	if (first_out) {
		cout << root;
		first_out = false;
	}
	else {
		cout << " " << root;
	}
	inorder_invert(Tree[root].left);
}

int main() {
	
	int exist_node[10] = { 0 };
	int N;
	cin >> N;
	for (int i = 0; i < N; i++) {
		char left, right;
		cin >> left >> right;
		Tree[i].left = transfer(left);
		Tree[i].right = transfer(right);
		node n = Tree[i];

		if (n.left != -1) exist_node[n.left]++;
		if (n.right != -1) exist_node[n.right]++;
	}
	int root = 0;		
	//	找出root节点，即孩子节点中没出现过的节点

	for (root; root < N; root++) {
		if (exist_node[root] == 0) {
			break;
		}
	}
	level_invert_order(root);
	cout << endl;
	inorder_invert(root);
	return 0;
}

File: pat1102.py
"""
    思路：1.先找到根节点，再通过根节点建立二叉树
          2.通过反转层序遍历和反转中序遍历输出结果
"""

class node():
    
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

N = int(input())
node_dic = {}           #   记录节点之间的对应关系
nodes = [0] * N         #   用于判断节点有没有出现
level = []              #   用于层序遍历的队列
inorder_result = []      #   用于记录中序反转序列结果

#   建立二叉树
def create_BTree(root):
    children = node_dic[root.value]
    if children[0] == '-' and children[1] == '-':
        return 
    if children[0] != '-':
        root.left = node(children[0])
        create_BTree(root.left)
    if children[1] != '-':
        root.right = node(children[1])
        create_BTree(root.right)

#   反转层序遍历
def invert_level_order(root):
    level.append(root)
    print(root.value, end='')
    while len(level) != 0:
        pop_node = level.pop(0)
        if pop_node.right is not None:
            level.append(pop_node.right)
            print(" " + pop_node.right.value, end='')
        if pop_node.left is not None:
            level.append(pop_node.left)
            print(" " + pop_node.left.value, end='')

#   反转中序遍历
def inorder_invert(root):
    if root is None:
        return
    else:
        inorder_invert(root.right)
        inorder_result.append(root.value)
        inorder_invert(root.left)


for x in range(N):
    child_list = input().split()
    #   记录出现过的节点
    for child in child_list:
        if child != '-':
            nodes[int(child)] += 1
    node_dic[str(x)] = child_list

#   没出现过的节点即为根节点
for i,v in enumerate(nodes):
    if v == 0:
        break
root = node(str(i))
create_BTree(root)
invert_level_order(root)
inorder_invert(root)
print()
print(inorder_result[0], end='')
for x in inorder_result[1:]:
    print(" " + x, end='')





